sdks/python/apache_beam/runners/portability/fn_api_runner_execution.py

```diff
@@ -0,0 +1,369 @@
+        windowing_strategy = pipeline_context.windowing_strategies[
+            pipeline_components.pcollections[
+                output_pcoll].windowing_strategy_id]
+        pcoll_buffers[buffer_id] = _GroupingBuffer(
+            pre_gbk_coder, post_gbk_coder, windowing_strategy)
+    else:
+      # These should be the only two identifiers we produce for now,
+      # but special side input writes may go here.
+      raise NotImplementedError(buffer_id)
+    return pcoll_buffers[buffer_id]
+  return buffer_fetcher
+
+
+def get_input_operation_name(
+    process_bundle_descriptor, transform_id, input_id):
+  # type: (beam_fn_api_pb2.ProcessBundleDescriptor, str, str) -> str
+  """Returns a callable to find the ID of the data input operation that
+  feeds an input PCollection to a PTransform. """
+  input_pcoll = process_bundle_descriptor.transforms[
+      transform_id].inputs[input_id]
+  for read_id, proto in process_bundle_descriptor.transforms.items():
+    if (proto.spec.urn == bundle_processor.DATA_INPUT_URN
+        and input_pcoll in proto.outputs.values()):
+      return read_id
+  raise RuntimeError(
+      'No IO transform feeds %s' % transform_id)
+
+
+def store_side_inputs_in_state(
+    worker_handler,  # type: WorkerHandler
+    context,  # type: pipeline_context.PipelineContext
+    pipeline_components,  # type: beam_runner_api_pb2.Components
+    data_side_input,  # type: DataSideInput
+    pcoll_buffers,  # type: Mapping[bytes, PartitionableBuffer]
+    safe_coders
+):
+  # type: (...) -> None
+  for (transform_id, tag), (buffer_id, si) in data_side_input.items():
+    _, pcoll_id = split_buffer_id(buffer_id)
+    value_coder = context.coders[safe_coders[
+        pipeline_components.pcollections[pcoll_id].coder_id]]
+    elements_by_window = _WindowGroupingBuffer(si, value_coder)
+    for element_data in pcoll_buffers[buffer_id]:
+      elements_by_window.append(element_data)
+
+    if si.urn == common_urns.side_inputs.ITERABLE.urn:
+      for _, window, elements_data in elements_by_window.encoded_items():
+        state_key = beam_fn_api_pb2.StateKey(
+            iterable_side_input=beam_fn_api_pb2.StateKey.IterableSideInput(
+                transform_id=transform_id,
+                side_input_id=tag,
+                window=window))
+        worker_handler.state.append_raw(state_key, elements_data)
+    elif si.urn == common_urns.side_inputs.MULTIMAP.urn:
+      for key, window, elements_data in elements_by_window.encoded_items():
+        state_key = beam_fn_api_pb2.StateKey(
+            multimap_side_input=beam_fn_api_pb2.StateKey.MultimapSideInput(
+                transform_id=transform_id,
+                side_input_id=tag,
+                window=window,
+                key=key))
+        worker_handler.state.append_raw(state_key, elements_data)
+    else:
+      raise ValueError(
+          "Unknown access pattern: '%s'" % si.urn)
+
+def add_residuals_and_channel_splits_to_deferred_inputs(
```

Comment:
This is one of those methods that doesn't make very much sense out of context. At least add a docstring?